Mathlib/Data/PNat/Factors.lean

```diff
@@ -0,0 +1,428 @@
+#align prime_multiset.prod_smul PrimeMultiset.prod_smul
+
+end PrimeMultiset
+
+namespace PNat
+
+/-- The prime factors of n, regarded as a multiset -/
+def factorMultiset (n : ℕ+) : PrimeMultiset :=
+  PrimeMultiset.ofNatList (Nat.factors n) (@Nat.prime_of_mem_factors n)
+#align pnat.factor_multiset PNat.factorMultiset
+
+/-- The product of the factors is the original number -/
+theorem prod_factorMultiset (n : ℕ+) : (factorMultiset n).prod = n :=
+  eq <| by
+    dsimp [factorMultiset]
+    rw [PrimeMultiset.prod_ofNatList]
+    exact Nat.prod_factors n.ne_zero
+#align pnat.prod_factor_multiset PNat.prod_factorMultiset
+
+theorem coeNat_factorMultiset (n : ℕ+) :
+    (factorMultiset n : Multiset ℕ) = (Nat.factors n : Multiset ℕ) :=
+  PrimeMultiset.to_ofNatMultiset (Nat.factors n) (@Nat.prime_of_mem_factors n)
+#align pnat.coe_nat_factor_multiset PNat.coeNat_factorMultiset
+
+end PNat
+
+namespace PrimeMultiset
+
+/-- If we start with a multiset of primes, take the product and
+ then factor it, we get back the original multiset. -/
+theorem factorMultiset_prod (v : PrimeMultiset) : v.prod.factorMultiset = v := by
+  apply PrimeMultiset.coeNat_injective
+  suffices toNatMultiset (PNat.factorMultiset (prod v)) = toNatMultiset v by exact this
+  rw [v.prod.coeNat_factorMultiset, PrimeMultiset.coe_prod]
+  rcases v with ⟨l⟩
+  --unfold_coes
+  dsimp [PrimeMultiset.toNatMultiset]
+  rw [Multiset.coe_prod]
+  let l' := l.map (Coe.coe : Nat.Primes → ℕ)
+  have : ∀ p : ℕ, p ∈ l' → p.Prime := fun p hp => by
+    rcases List.mem_map.mp hp with ⟨⟨_, hp'⟩, ⟨_, h_eq⟩⟩
+    exact h_eq ▸ hp'
+  exact Multiset.coe_eq_coe.mpr (@Nat.factors_unique _ l' rfl this).symm
+#align prime_multiset.factor_multiset_prod PrimeMultiset.factorMultiset_prod
+
+end PrimeMultiset
+
+namespace PNat
+
+/-- Positive integers biject with multisets of primes. -/
+def factorMultisetEquiv : ℕ+ ≃ PrimeMultiset
+    where
+  toFun := factorMultiset
+  invFun := PrimeMultiset.prod
+  left_inv := prod_factorMultiset
+  right_inv := PrimeMultiset.factorMultiset_prod
+#align pnat.factor_multiset_equiv PNat.factorMultisetEquiv
+
+/-- Factoring gives a homomorphism from the multiplicative
+ monoid ℕ+ to the additive monoid of multisets. -/
+theorem factorMultiset_one : factorMultiset 1 = 0 := by
+  simp [factorMultiset, PrimeMultiset.ofNatList, PrimeMultiset.ofNatMultiset]
+#align pnat.factor_multiset_one PNat.factorMultiset_one
+
+theorem factorMultiset_mul (n m : ℕ+) :
+    factorMultiset (n * m) = factorMultiset n + factorMultiset m := by
+  let u := factorMultiset n
+  let v := factorMultiset m
+  have : n = u.prod := (prod_factorMultiset n).symm; rw [this]
+  have : m = v.prod := (prod_factorMultiset m).symm; rw [this]
+  rw [← PrimeMultiset.prod_add]
+  repeat' rw [PrimeMultiset.factorMultiset_prod]
+#align pnat.factor_multiset_mul PNat.factorMultiset_mul
+
+theorem factorMultiset_pow (n : ℕ+) (m : ℕ) : factorMultiset (n ^ m) = m • factorMultiset n := by
```

Comment:
Lean doesn't seem to find the `Pow ℕ+ ℕ` instance, and insists on using the `Pow ℕ ℕ` instance (inserting a coercion). Not sure how to solve that.

Reply:
Made it compile by replacing  `^` with `Pow.pow`, this also reproduces the original meanings of theorems from mathlib3, like in `prod_smul`. Zulip thread for further discussion: [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EPNat.2EFactors.20!4.231830)